### DISCUSSION TITLE
Is it possible to infer types for `add_task` via `typing.ParamSpec`?

### DISCUSSION THREAD

**Discussion by jd-solanki** [encode/starlette#2694] [Ideas] [answered]
Hi 👋🏻 

Great fan of encode ❤️ 

As python type system is improving can we use [`typing.ParamSpec`](https://docs.python.org/3/library/typing.html#typing.ParamSpec) somehow to infer the type for arguments we need to pass to `add_task` callable?

Right now, If we write this: `.add_task(my_task, |)` then we don't get any type hints at `|` about what arguments `my_task` accepts.

I'm not type wizard but if `typing.ParamSpec` can infer types of callback (first arg) then it'll nice addition.

Thanks 😇

**Comment by Kludex** [comment, marked as answer]
We do use it... https://github.com/encode/starlette/blob/master/starlette/background.py

**Reply by jd-solanki** [reply]
I was trying it in FastAPI and main goal was to get the autocompletion for the params but it looks like we aren't getting that is that expected?

![image](https://github.com/user-attachments/assets/ed5edad6-0dfe-489f-a93d-d49b9c3040e8)

```py
import time

from fastapi import BackgroundTasks, FastAPI

app = FastAPI()


def background_task(*, message: str):
    time.sleep(3)
    print(message)


@app.get("/")
async def read_root(background_tasks: BackgroundTasks):
    background_tasks.add_task(background_task, message="Hello, World!")
    return {"message": "Hello, World!"}
```

I think this will be same with starlette 🤔

**Reply by jd-solanki** [reply]
Here's starlette example:

![image](https://github.com/user-attachments/assets/fb844cb6-e63b-440e-8dd7-e429625a8d81)


```py
import time

from starlette.applications import Starlette
from starlette.background import BackgroundTask
from starlette.requests import Request
from starlette.responses import JSONResponse
from starlette.routing import Route


def background_task(*, message: str):
    time.sleep(3)
    print(message)


async def homepage(request: Request) -> JSONResponse:
    task = BackgroundTask(background_task, message="hello")
    return JSONResponse({"hello": "world"}, background=task)


app = Starlette(
    debug=True,
    routes=[
        Route("/", homepage),
    ],
)
```

> _P.S. You been to india? when?_

**Reply by Kludex** [reply]
<img width="634" alt="Screenshot 2024-09-24 at 08 58 22" src="https://github.com/user-attachments/assets/0715a47c-4e9a-4592-8f9b-3c0cb214cd42">

It works here.

> P.S. You been to india? when?

Yep, last year. I was invited to give a talk in Hyderabad. 🙏

**Reply by jd-solanki** [reply]
Yes, We're getting type error in my example as well so that's working.

Can we make it work with autocompletion as well?

**Reply by Kludex** [reply]
I think that's on the IDEs/language server's side...
